qldpc/abstract.py

```diff
@@ -525,3 +598,110 @@ def lift(member: int) -> IntegerArray:
+    """Projective variant of the special linear group."""
+
+    def __init__(self, dimension: int, field: int | None = None) -> None:
+        base_field = galois.GF(field or DEFAULT_FIELD_ORDER)
+        if base_field.order == 2:
+            super().__init__(SpecialLinearGroup(dimension, 2))
+        elif dimension == 2:
+            generators = self.get_expanding_generator_mats(dimension, field)
+            space = _construct_projspace(dimension, field)
+            group = Group.from_permutation_mats(generators, space, field)
+            super().__init__(group)
+        else:
+            raise ValueError(
+                "Projective special linear groups with both dimension and field greater than 2 are"
+                " not yet supported."
+            )
+
+    @classmethod
+    def get_expanding_generator_mats(
+        cls, field: int | None = None
+    ) -> tuple[galois.FieldArray, galois.FieldArray]:
+        """Expanding generator matrices for PSL(2, field), from https://arxiv.org/abs/1807.03879"""
+        base_field = galois.GF(field or DEFAULT_FIELD_ORDER)
+        minus_one = -base_field(1)
+        A = base_field([[1, 1], [0, 1]])
+        B = base_field([[1, minus_one], [0, 1]])
+        C = base_field([[1, 0], [1, 1]])
+        D = base_field([[1, 0], [minus_one, 1]])
+        return A, B, C, D
+
+
+def _construct_linspace(dimension: int, field: int | None = None) -> list[bytes]:
+    """Helper function to generate a list of vectors over finite field."""
+    base_field = galois.GF(field or DEFAULT_FIELD_ORDER)
+    vectors = itertools.product(base_field.elements, repeat=dimension)
+    next(vectors)  # skip the all-0 element
+    return [base_field(vec).tobytes() for vec in vectors]
+
+
+def _construct_projspace(dimension: int, field: int) -> list[bytes]:
+    """Helper function to create the vectors in the projective space.
+    The difference from usual vectors is that scalar multiples are identified.
+    """
+    base_field = galois.GF(field)
+    return [
+        base_field(vec).tobytes()
+        for vec in itertools.product(base_field.elements, repeat=dimension)
+        if vec[(base_field(vec) != 0).argmax()] == 1
+    ]
+
+
+def construct_special_linear_groups(
+    dimension: int, field: int | None = None
+) -> tuple[list[galois.FieldArray], list[galois.FieldArray]]:
+    """Construct all elements of SL(dimension, field) and PSL(field, dimension) by brute force.
+
+    WARNING: SLOW!!!
+    """
+    special_linear: list[galois.FieldArray] = []
+    proj_special_linear: list[galois.FieldArray] = []
+    base_field = galois.GF(field or DEFAULT_FIELD_ORDER)
+    for entries in itertools.product(base_field.elements, repeat=dimension**2):
+        vec = base_field(entries)
+        mat = vec.reshape(dimension, dimension)
+        if np.linalg.det(mat) == 1:
+            special_linear.append(mat)  # type:ignore[arg-type]
+            # for PSL, we force the first non-zero entry to be < p/2 to quotient by -I
```

Comment:
I trust you on this, but I don't quite understand it due to my limited understanding of finite groups.  Let's discuss.

Reply:
Yeah sure. PSL is basically SL but scalar multiples (by d^th roots of unity) are treated as same. So, for d =2, one has to pick only one among {A, -A}. One way to do that is to force a "standard form," i.e., exactly one among {A, -A} has the first entry < p/2. Hence, the rule.

Reply:
Alright that was simpler than I anticipated :sweat_smile: 

Makes sense. Thanks!